### ISSUE TITLE
Allow scopes to be set in a config file

### ISSUE THREAD

**Issue by sdondley** [sdondley/moo-google#3]
Allow users to modify a scopes.cfg file that they can use to set the scopes for authorization.

**Comment by pscott-au** [comment]
I think you'll find the fixes in my repo at https://github.com/pscott-au/moo-google/commit/7b86ddaf934662b0d66e538947877140f1faa393 .. possibly may be a couple of tweaks either side

**Comment by sdondley** [comment]
ok, but I need clean commits. You still have readme file changes mixed in with it. makes it much more confusing and harder to merge in.

**Comment by pscott-au** [comment]
Just a quick FYI - I committed a bug earlier - a typo into
WebService::Google::Client::Discovery

my @doclinks = uniq mxap { $_->{documentationLink} } grep { $_->{name} eq
$s } @$all;

Should be

my @doclinks = uniq map { $_->{documentationLink} } grep { $_->{name} eq $s
} @$all;

BUG introduced in "Mostly style refactor"
https://github.com/pscott-au/moo-google/commit/7bc2a4ecd229f1029e0c44ba324c376a18fa4904#diff-3ed2e98270882de829098a9aa66ae5ae
15 hrs ago

This was breaking my previously working ACL test  - so yes - i will take a
little time to create some tests that make sense to run before commits. I
don't like his test but with the minor changes I've made to the code at
least is possible to make them configurable and not require as much hard
coded stuff.


RE the README.md issue - I'm pretty sure that my most recent code is
drawing this from the PM code but easy fix is to run the test and commit
the diff back in if you want to keep my history. If you don't want my code
that is ok but I'm not going to do another branch to revert to my pull
request then change the README which gets overwritten on build anyway.

**Comment by pscott-au** [comment]
A side issue re your DZilla dist - I checked it out and was going to do a
pull request to put either page1 or create an index page as the README.md
so that it shows when ppl browse the repo. Prolly easier if I just suggest
it and let you decide how to handle that one.

**Comment by pscott-au** [comment]
just as an FYI - I'm exposing my tarball at https://pscott-au.github.io/WebService-Google/WebService-Google-Client-0.05.tar.gz which includes the fix for this amongst others. NB I've moved to using gapi.json as the default config and allow the scope to be set on file creation with goauth from the command line.

**Comment by sdondley** [comment]
OK, thanks. How is gapi.json different than config.json and how is it generated?

**Comment by pscott-au** [comment]
Is all the same as config.json just a more meaningful filename that is less
likely to be clobbered or conflict with other systems.

**Comment by sdondley** [comment]
Ah, gotcha. Do you know what google or OAuth 2.0 calls the payload in that file? Maybe we should name it after that.

**Comment by pscott-au** [comment]
It's a structure describing the Access Tokens .. I think trying to bind it to something like GoogleOATH2JWT or something is more problematic that it's worth, but may be worth looking at how the many other Google perl libs describe, contain and store it. eg Mojo::JWT::Google

See also #19 which includes code that resolves this issue.

**Comment by sdondley** [comment]
Looks like "json key file" is one possibility, according to this tutorial: https://flaviocopes.com/google-api-authentication/. No idea if that's official. Maybe they made it up, too.

My larger point is we should try to use established terminology to help developers avoid confusion.

**Comment by pscott-au** [comment]
yes I agree - but care is needed. I chose gapi.json as opposed to
config.json for this very reason. Need to keep in mind that this file
contains a few distinct components. It contains user access and refresh
tokens as well as describing the service account. Each component has well
defined definitions that you could align with across different audiences
such as the Google Docs but collectively it's not quite so simple.

**Comment by sdondley** [comment]
Well, gapi.conf is an improvement, I'm just wondering if there is a more technically accurate name for it.

**Comment by pscott-au** [comment]
gapi.json

**Comment by sdondley** [comment]
According to what authority?

**Comment by pscott-au** [comment]
none - it's a json file so a json extension would seem to be to be appropriate - of course if there is an authority that states that conf is the better extension post it here and I'm happy to review it.

**Comment by sdondley** [comment]
It was config.json and you have changed it to gapi.json. I'm more curious about the "gapi" part. When I see that, I think "Google API." But really, what's in the file is not an API, but the data that lets me access data via that API.

**Comment by pscott-au** [comment]
ok - was a trivial matter but by all means call it google_auth_credentials.json or whatever strikes you as appropriate - I'd argue that gapi is short and concise and descriptive and that json is a suitable extension. It is user configurable through CLI so if ppl disagree there is an option. There is no officially supported way to describe the file that contains your google access project auth definition including user access and refresh tokens to the best of my knowledge.

**Comment by pscott-au** [comment]
config.json just struck me as problematic
